src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1591,119 +1566,184 @@ static size_t delete_monitors(GrowableArray<ObjectMonitor*>* delete_list) {
-      HandshakeForDeflation hfd_hc;
-      Handshake::execute(&hfd_hc);
-      // Also, we sync and desync GC threads around the handshake, so that they can
-      // safely read the mark-word and look-through to the object-monitor, without
-      // being afraid that the object-monitor is going away.
-      VM_RendezvousGCThreads sync_gc;
-      VMThread::execute(&sync_gc);
-
-      if (ls != nullptr) {
-        ls->print_cr("after handshaking: in_use_list stats: ceiling="
-                     SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
-                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
-        timer.start();
-      }
-    } else {
-      // This is not a monitor deflation thread.
-      // No handshake or rendezvous is needed when we are already at safepoint.
-      assert_at_safepoint();
-    }
+    log.before_handshake(unlinked_count);
+
+    // A JavaThread needs to handshake in order to safely free the
+    // ObjectMonitors that were deflated in this cycle.
+    HandshakeForDeflation hfd_hc;
+    Handshake::execute(&hfd_hc);
+    // Also, we sync and desync GC threads around the handshake, so that they can
+    // safely read the mark-word and look-through to the object-monitor, without
+    // being afraid that the object-monitor is going away.
+    VM_RendezvousGCThreads sync_gc;
+    VMThread::execute(&sync_gc);
+
+    log.after_handshake();
 
     // After the handshake, safely free the ObjectMonitors that were
     // deflated and unlinked in this cycle.
-    if (current->is_Java_thread()) {
-      if (ls != NULL) {
-        timer.stop();
-        ls->print_cr("before setting blocked: unlinked_count=" SIZE_FORMAT
-                     ", in_use_list stats: ceiling=" SIZE_FORMAT ", count="
-                     SIZE_FORMAT ", max=" SIZE_FORMAT,
-                     unlinked_count, in_use_list_ceiling(),
-                     _in_use_list.count(), _in_use_list.max());
-      }
-      // Mark the calling JavaThread blocked (safepoint safe) while we free
-      // the ObjectMonitors so we don't delay safepoints whilst doing that.
-      ThreadBlockInVM tbivm(JavaThread::cast(current));
-      if (ls != NULL) {
-        ls->print_cr("after setting blocked: in_use_list stats: ceiling="
-                     SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
-                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
-        timer.start();
-      }
-      deleted_count = delete_monitors(&delete_list);
-      // ThreadBlockInVM is destroyed here
-    } else {
-      // A non-JavaThread can just free the ObjectMonitors:
-      deleted_count = delete_monitors(&delete_list);
-    }
+
+    log.before_blocked(unlinked_count);
+
+    // Mark the calling JavaThread blocked (safepoint safe) while we free
+    // the ObjectMonitors so we don't delay safepoints whilst doing that.
+    ThreadBlockInVM tbivm(current);
+
+    log.after_blocked();
```

Comment:
This sequence is pre-existent but the check for safepoint/handshakes happens in ~ThreadBlockInVM so there is no point in logging before and after ThreadBlockInVM since we are not going to block. Also with this sequence the timer will still run if we do block in ~ThreadBlockInVM. So I would remove them and move log.end() at the end of this block before the ~ThreadBlockInVM runs since we are done by then. If we want to log when we the deflater actually returns then we can add a log there too.

Reply:
Good point. This code is going to be rewritten for https://bugs.openjdk.org/browse/JDK-8319137 and https://bugs.openjdk.org/browse/JDK-8319048. The comment here https://github.com/openjdk/jdk/pull/16412#issuecomment-1818358260 is a suggestion for how to solve both issues, and I think it should include the fixes to the logging.

Reply:
The call to `log.before_blocked()` on L1725 will stop the timer so it will not
be running during the ThreadBlockInVM.
The call to `log.after_blocked()` on L1731 will restart the timer.
The advantage of having these two calls is that you know how long we spent
in ThreadBlockInVM based on the timestamps in the two log mesgs.

Reply:
Definitely going to have to merge with [JDK-8319137](https://bugs.openjdk.org/browse/JDK-8319137)!!

Reply:
>The call to log.before_blocked() on L1725 will stop the timer so it will not
be running during the ThreadBlockInVM.
The call to log.after_blocked() on L1731 will restart the timer.
The advantage of having these two calls is that you know how long we spent in ThreadBlockInVM based on the timestamps in the two log mesgs.

It's unclear to me if you want me to change anything. Patricio pointed out that the pre-existing logs are around the "wrong" part of the code. It's not interesting to log the constructor, but rather the destructor. Alternatively, the code that deletes the monitors.

Reply:
I missed Patricio's point that the "work" time is done in the destructor and not in the
constructor. So my original logging was not useful (but it was fast!! :-)) and didn't
achieve the desired result.

Reply:
When the dust settles with the Bug fixes in this area, we should take a look and make sure we have the correct logging we want.